### PR TITLE
Connector dataset dropdown

### PR DESCRIPTION
**Pull request by allisonking**
Closes https://github.com/ethyca/fides/issues/2114

### Code Changes

* [x] Adds a dataset dropdown next to the existing yaml editor
* [x] Allows the dataset dropdown to link a dataset to a connector config
* [x] Additional confirmation button for overwriting datasets in the yaml editor
* [x] Fix copy (system yaml --> dataset YAML)
* [x] Cypress tests

### Steps to Confirm

* [ ] `nox -s test_env`
* [ ] Configure a connector dataset

### Pre-Merge Checklist

* [ ] All CI Pipelines Succeeded
* Documentation:
  * [ ] documentation complete, [PR opened in fidesdocs](https://github.com/ethyca/fidesdocs/pulls)
  * [ ] documentation [issue created in fidesdocs](https://github.com/ethyca/fidesdocs/issues/new/choose)
* [ ] Issue Requirements are Met
* [ ] Relevant Follow-Up Issues Created
* [x] Update `CHANGELOG.md`

### Description Of Changes

Using the dropdown

https://user-images.githubusercontent.com/24641006/211687577-d15544a6-4ed1-4fd8-b27b-91840857f52b.mov

Using the yaml editor

https://user-images.githubusercontent.com/24641006/211687670-72ddcd58-c9b9-4319-b864-659f38f8e692.mov

### CODE REVIEW

**Comment by allisonking** [comment]
Failing test seems unrelated?

**Comment by pattisdr** [comment]
yep this is unrelated @allisonking, I saw this on my last `unified-fides-resources` PR but I don't think it's on main. After we merge your branch in, I'll get `unified-fides-resources` up-to-date with main and investigate there.

**Review by pattisdr — CHANGES_REQUESTED** [review]
This feels a little buggy to me in places. This branch isn't up-to-date with `unified-fides-resources` that has the [improved upsert validation](https://github.com/ethyca/fides/pull/2134), but that is handy here so it's easy to see where we might add better error handling: 


1) Start from scratch, delete datasetconfigs, connectionconfigs, and ctl_datasets. Refresh.  It makes sense that there's no dropdown to connect my first connection config to an existing dataset, but the save button stacked on the cancel button looks off
Screen Shot 2023-01-11 at 12 39 39 PM
2)  Copy and paste a dataset (mysql is a good one) that doesn't have an `organization_fides_key` or `data_qualifier` in it.  You could also send in a bad data category.  We get failed requests but the UI looks like we never submitted a request at all. (In the background a ctl_dataset has been created, but the ops datasetconfig was not created)

https://user-images.githubusercontent.com/9755598/211892518-7d8c0b07-ce39-4262-a2c9-f6b0159d6cac.mov

3) Editing an existing dataset that was good to something invalid (add bad data category) also puts the UI in a weird state
Screen Shot 2023-01-11 at 12 57 39 PM

4) When saving existing dataset yaml you've added a helpful confirmation that you're about to overwrite a dataset.  This is not necessarily true if the user is also editing the `fides_key`.
Below, the only edit I made was to change `postgres_example_test_dataset` fides_key to `updated_fides_key`.  This will create a new `ctl_dataset` titled `updated_fides_key`, and link it to the DatasetConfig.


Screen Shot 2023-01-11 at 1 03 11 PM


5) When configuring an existing dataset in the connector workflow, it took me awhile to figure out how to edit directly.  I needed to unselect the current dataset.

**Comment by allisonking** [comment]
Thanks for the thorough testing @pattisdr ! 🙏 

1. This should be fixed. It's a little weird code-wise, but the new designs don't have a cancel button anymore so I don't think we have to worry too much about it since it'll go away soon.
image

2 and 3. I'm having trouble reproducing these—I just merged in `unified-fides-resources` which like you said throws the proper errors now, so I wonder if that's letting me see good errors instead of getting stuck in weird states? Can you let me know if you're still seeing these problems, and if so, with what dataset? (I tried out `mysql_example_test_dataset.yml` which seemed to work ok)

4. Good catch, should be fixed!
5. Yeah, this was the decision we came to when we were trying to decide on behavior without design. The good news is this will go away with the new design that's currently in the works

File: clients/admin-ui/src/features/datastore-connections/add-connection/forms/YamlEditorForm.tsx
Comment (pattisdr) [inline clients/admin-ui/src/features/datastore-connections/add-connection/forms/YamlEditorForm.tsx:101]: I think the check is really does this `fides_key` match any fides_key in the ctl_datasets table? Because we could still update it to match a different ctl_dataset and that would be even more unintuitive. 

Regardless, if you update the key, it now looks like you don't have an option to save anything at all, you just click the save button and nothing happens.   It's probably safer if they can't update the fides_key without a warning, but the UI is in a weird state still.

**Comment by pattisdr** [comment]
2 and 3. I'm having trouble reproducing these—I just merged in unified-fides-resources which like you said throws the proper errors now, so I wonder if that's letting me see good errors instead of getting stuck in weird states? Can you let me know if you're still seeing these problems, and if so, with what dataset? (I tried out mysql_example_test_dataset.yml which seemed to work ok)


OK after some more digging I found what bad request is causing this, it's if a `GET http://localhost:3000/api/v1/dataset` throws an error because it's returning a bad dataset.   It's harder to create a bad dataset after you've merged in `unified-fides-resources`.  

To reproduce, you could force the backend to throw an error, for example:
1) Go to src/fides/api/ctl/routes/crud.py
2) In the list endpoint `async def ls`:, raise an exception first thing Exception()

This mimics what would happen if a user had a bad dataset already saved.

File: clients/admin-ui/src/features/datastore-connections/add-connection/forms/YamlEditorForm.tsx
Comment (pattisdr) [inline clients/admin-ui/src/features/datastore-connections/add-connection/forms/YamlEditorForm.tsx:140]: Bravo this looks great

File: clients/admin-ui/src/features/datastore-connections/add-connection/forms/YamlEditorForm.tsx
Comment (allisonking) [reply clients/admin-ui/src/features/datastore-connections/add-connection/forms/YamlEditorForm.tsx:101]: ah I see, okay yes there are two things going on here, thanks for catching that!

**Comment by pattisdr** [comment]
wonderful, testing now!

File: clients/admin-ui/src/features/datastore-connections/add-connection/DatasetConfiguration.tsx
Comment (pattisdr) [inline clients/admin-ui/src/features/datastore-connections/add-connection/DatasetConfiguration.tsx:167]: OK great improvements if there are errors loading bad datasets!

I didn't really notice the line of text though: `There was a problem loading existing datasets, please try again.` as it didn't stand out, it's not clear there was an error.  If you're going to add text directly to the page, could it be red, or perhaps a toast message instead?
Screen Shot 2023-01-12 at 11 07 40 AM

File: clients/admin-ui/src/features/datastore-connections/add-connection/forms/YamlEditorForm.tsx
Comment (pattisdr) [inline clients/admin-ui/src/features/datastore-connections/add-connection/forms/YamlEditorForm.tsx:103]: The message in the warning modal may not always be technically correct: the ctl_dataset that we're about to upsert, is going to match whatever the `fides_key` is in the first line of the yaml file. 

So here, if the dataset was originally `my_other_mongo_test` and I have just pasted yaml that has `mongo_test`, it is `mongo_test` that is going to be either created or updated. 

Screen Shot 2023-01-12 at 11 12 43 AM

**Review by pattisdr — COMMENTED** [review]
OK this UI is feeling pretty good now and holding up under a lot of scenarios I'm testing. Thanks for your nice changes here! There's just a couple of things I saw on this last round but I think it's about ready to go.

File: clients/admin-ui/src/features/datastore-connections/add-connection/DatasetConfiguration.tsx
Comment (allisonking) [reply clients/admin-ui/src/features/datastore-connections/add-connection/DatasetConfiguration.tsx:167]: sure thing!

image

**Comment by pattisdr** [comment]
@allisonking one more thing I just thought of I forgot to test: creating saas connectors.  One big difference with saas connectors is that we create them from a template.  After the user enters their connector parameters and clicks save, the connection config, dataset config, and dataset are all created in the background for them. 

There's a small thing where it's hard to save the dataset when you create multiple saas connectors.
- Create a mailchimp connector
- Enter connector parameters (just fake data is fine)
- On configure dataset click save in the dropdown
- Create a segment connector
- Enter connector parameters
- On configure dataset, the segment dataset is not in the dropdown.  I can click "Save" here, but it's confusing because there's only "Select" in the dropdown. I can't edit the yaml either: by toggling to another dataset, then back to Select, you can edit the yaml.  I suspect the "Select" is tied to the actual dataset here behind the scenes. 


https://user-images.githubusercontent.com/9755598/212152362-2ac27733-b698-444d-b2a5-7233d2091c2e.mov

**Comment by allisonking** [comment]
Oh man, that's a good catch I never would have looked for. [This should fix it](https://github.com/ethyca/fides/pull/2162/commits/19551b0cdb4a51da0969bcd12e48460b12c95450). The change really should have only been the one line (to add "Datasets" to the list of resources in the cache to be invalidated when a new saas connector is created), but [due to some tech debt](https://github.com/ethyca/fides/issues/1386) I had to combine the relevant redux slices to use the same base API first.

**Review by pattisdr — APPROVED** [review]
Alright this looks really good to me @allisonking thanks for your hard work here 🏆  Failing test is unrelated like we discussed and I'll look into this after you merge this branch into `unified-fides-resources`.

**Pull request merged**
